nativelib/src/main/resources/gc/commix/Heap.c

```diff
@@ -0,0 +1,542 @@
+    for (int i = 0; i < gcThreadCount; i++) {
+        Stats *stats;
+#ifdef ENABLE_GC_STATS
+        if (statsFile != NULL) {
+            int len = strlen(statsFile) + 5;
+            char *threadSpecificFile = (char *) malloc(len);
+            snprintf(threadSpecificFile, len, "%s.t%d", statsFile, i);
+            stats = malloc(sizeof(Stats));
+            Stats_Init(stats, threadSpecificFile, (uint8_t)i);
+        } else {
+#endif
+            stats = NULL;
+#ifdef ENABLE_GC_STATS
+        }
+#endif
+        GCThread_Init(&gcThreads[i], i, heap, stats);
+    }
+
+    heap->mark.lastEnd_ns = scalanative_nano_time();
+
+    pthread_mutex_init(&heap->sweep.growMutex, NULL);
+}
+
+word_t *Heap_AllocLarge(Heap *heap, uint32_t size) {
+
+    assert(size % ALLOCATION_ALIGNMENT == 0);
+    assert(size >= MIN_BLOCK_SIZE);
+
+    Object *object = LargeAllocator_GetBlock(&largeAllocator, size);
+    if (object != NULL) {
+done:
+        assert(object != NULL);
+        assert(Heap_IsWordInHeap(heap, (word_t *)object));
+        return (word_t *)object;
+}
+
+    if (!Sweeper_IsSweepDone(heap)) {
+        object = Sweeper_LazySweepLarge(heap, size);
+        if (object != NULL)
+            goto done;
+    }
+
+    Heap_Collect(heap);
+
+    object = LargeAllocator_GetBlock(&largeAllocator, size);
+    if (object != NULL)
+        goto done;
+
+    if (!Sweeper_IsSweepDone(heap)) {
+        object = Sweeper_LazySweepLarge(heap, size);
+        if (object != NULL)
+            goto done;
+    }
+
+    size_t increment = MathUtils_DivAndRoundUp(size, BLOCK_TOTAL_SIZE);
+    uint32_t pow2increment = 1U << MathUtils_Log2Ceil(increment);
+    Heap_Grow(heap, pow2increment);
+
+    object = LargeAllocator_GetBlock(&largeAllocator, size);
+
+    goto done;
+}
+
+NOINLINE word_t *Heap_allocSmallSlow(Heap *heap, uint32_t size) {
+    Object *object = (Object *) Allocator_Alloc(&allocator, size);
+
+    if (object != NULL) {
+done:
+        assert(Heap_IsWordInHeap(heap, (word_t *)object));
+        assert(object != NULL);
+        ObjectMeta *objectMeta = Bytemap_Get(allocator.bytemap, (word_t *)object);
+#ifdef DEBUG_ASSERT
+        ObjectMeta_AssertIsValidAllocation(objectMeta, size);
+#endif
+        ObjectMeta_SetAllocated(objectMeta);
+        return (word_t *)object;
+    }
+
+    if (!Sweeper_IsSweepDone(heap)) {
+        object = Sweeper_LazySweep(heap, size);
+
+        if (object != NULL)
+            goto done;
+    }
+
+    Heap_Collect(heap);
+    object = (Object *) Allocator_Alloc(&allocator, size);
+
+    if (object != NULL)
+        goto done;
+
+    if (!Sweeper_IsSweepDone(heap)) {
+        object = Sweeper_LazySweep(heap, size);
+
+        if (object != NULL)
+            goto done;
+    }
+
+    // A small object can always fit in a single free block
+    // because it is no larger than 8K while the block is 32K.
+    Heap_Grow(heap, 1);
+    object = (Object *) Allocator_Alloc(&allocator, size);
```

Comment:
The two lines above look wrong. `Heap_Collect` will grow heap appropriately after collection so that 50% of the heap is free blocks. We have a single mutator, so there is no chance of someone else allocating all of the 50% of the heap before we reach here. 

In the hypothetical case of multiple mutator threads, it's better to do another collection instead of ad-hoc growth here. Because it looks like it might cause unbounded growth to the heap limit if this path is actually taken frequently.

Reply:
This still could happen if the reserved space (`SWEEP_RESERVE_BLOCKS`) is exactly as large as the free space after heap growth. When sweep is done, GC master thread takes the reserved blocks out of the free blocks. There is a race condition between this and the mutator allocating. If the GC thread wins and there are no recycled blocks, then the small allocation must grow the heap.